pallets/fees/src/lib.rs

```diff
@@ -83,106 +102,82 @@ pub mod pallet {
+			for (key, fee) in self.initial_fees.iter() {
+				<FeeBalances<T>>::insert(key, fee);
 			}
 		}
 	}
 
-	/// Stores the Fees associated with a Hash identifier
+	/// Stores the fee balances associated with a Hash identifier
 	#[pallet::storage]
 	#[pallet::getter(fn fee)]
-	pub(super) type Fees<T: Config> =
-		StorageMap<_, Blake2_256, T::Hash, Fee<T::Hash, BalanceOf<T>>>;
+	pub(super) type FeeBalances<T: Config> =
+		StorageMap<_, Blake2_256, T::FeeKey, BalanceOf<T>, ValueQuery, T::DefaultFeeValue>;
 
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
 	pub enum Event<T: Config> {
-		FeeChanged(T::Hash, BalanceOf<T>),
-	}
-
-	#[pallet::error]
-	pub enum Error<T> {
-		/// Fee associated to given key not found
-		FeeNotFoundForKey,
-		/// Invalid author of blocked fetched
-		InvalidAuthor,
+		FeeChanged(T::FeeKey, BalanceOf<T>),
 	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		/// Set the given fee for the key
 		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_fee())]
-		pub fn set_fee(
-			origin: OriginFor<T>,
-			key: T::Hash,
-			new_price: BalanceOf<T>,
-		) -> DispatchResult {
-			Self::can_change_fee(origin)?;
-			Self::change_fee(key, new_price);
-			Self::deposit_event(Event::FeeChanged(key, new_price));
+		pub fn set_fee(origin: OriginFor<T>, key: T::FeeKey, fee: BalanceOf<T>) -> DispatchResult {
+			T::FeeChangeOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)?;
+
+			<FeeBalances<T>>::insert(key, fee);
+			Self::deposit_event(Event::FeeChanged(key, fee));
 			Ok(())
 		}
 	}
 }
 
-impl<T: Config> Pallet<T> {
-	/// Called by any other module who wants to trigger a fee payment for a given account.
-	/// The current fee price can be retrieved via Fees::price_of()
-	pub fn pay_fee(from: T::AccountId, key: T::Hash) -> DispatchResult {
-		let fee = <Fees<T>>::get(key).ok_or(Error::<T>::FeeNotFoundForKey)?;
-		Self::pay_fee_to_author(from, fee.price)?;
-		Ok(())
-	}
-
-	/// Burns Fee from account
-	pub fn burn_fee(from: &T::AccountId, fee: BalanceOf<T>) -> DispatchResult {
-		let _ = T::Currency::withdraw(
-			from,
-			fee,
-			WithdrawReasons::FEE.into(),
-			ExistenceRequirement::KeepAlive,
-		)?;
+impl<T: Config, K: EncodeLike<T::FeeKey>> fees::Fees<K> for Pallet<T> {
+	type AccountId = T::AccountId;
+	type Balance = BalanceOf<T>;
 
-		Ok(())
+	fn fee_value(key: K) -> BalanceOf<T> {
+		<FeeBalances<T>>::get(key)
 	}
 
-	/// Pay the given fee
-	pub fn pay_fee_to_author(from: T::AccountId, fee: BalanceOf<T>) -> DispatchResult {
-		let value = T::Currency::withdraw(
-			&from,
-			fee,
-			WithdrawReasons::FEE.into(),
-			ExistenceRequirement::KeepAlive,
-		)?;
-
-		let author = <pallet_authorship::Pallet<T>>::author().ok_or(Error::<T>::InvalidAuthor)?;
-		T::Currency::resolve_creating(&author, value);
+	fn fee_to_author(from: &Self::AccountId, fee: Fee<BalanceOf<T>, K>) -> DispatchResult {
+		if let Some(author) = <pallet_authorship::Pallet<T>>::author() {
```

Comment:
Would it make sense to pass the pallet_authorship as trait?

Reply:
This idea is really cool, and I wanted to do the same, but the `pallet_authorship` do not exposes any trait for the `author()` method. I do not know if there is another approach to get this, if someone know, please shoot out!

Reply:
ah, I do not know if it is possible in the runtime definition to pass the function call instead? Otherwise nvm, I think is fine.

Reply:
You need a trait that gives that call. And then implement that it for the real authorship pallet somehow. Maybe the implementation of this new trait for the authorship pallet could be done in runtime/common. I like the idea, let me think about it, and probably for other PR it could be done.